io.openems.edge.io.revpi/src/io/openems/edge/io/revpi/RevPiDigitalIoDevice.java

```diff
@@ -164,44 +215,68 @@ public void handleEvent(Event event) {
 		}
 		switch (event.getTopic()) {
 		case EdgeEventConstants.TOPIC_CYCLE_BEFORE_PROCESS_IMAGE:
-			this.updateChannelValues();
+			this.eventBeforeProcessImage();
+			break;
+
+		case EdgeEventConstants.TOPIC_CYCLE_EXECUTE_WRITE:
+			this.eventExecuteWrite();
 			break;
 		}
 	}
 
+	/**
+	 * Execute on Cycle Event "Before Process Image".
+	 */
+	private void eventBeforeProcessImage() {
+		this.updateDataInChannels();
+	}
+
+	/**
+	 * Execute on Cycle Event "Execute Write".
+	 */
+	private void eventExecuteWrite() {
+		this.updateDataOutChannels();
+	}
+
+	/**
+	 * @brief reads the data either from the given DATA IN hardware port
+	 */
+	private boolean getData(int idx) throws IOException {
+		return this.revPiHardware.getDataIn(idx + 1);
+	}
+
+	private void appendBool(StringBuilder b, Optional<Boolean> val) {
+		if (val.isPresent()) {
+			if (val.get()) {
+				b.append("1");
+			} else {
+				b.append("0");
+			}
+		} else {
+			b.append("-");
+		}
+	}
+
 	@Override
 	public String debugLog() {
```

Comment:
In the KMTronic relay board implementation, we use a format like "xxx--xx" which i find easier to read for non-techies than "1110011". See https://github.com/OpenEMS/openems/blob/develop/io.openems.edge.io.kmtronic/src/io/openems/edge/io/kmtronic/KmtronicRelayOutput.java#L116 - but it's up to you.